src/sinks/gcp/cloud_storage.rs

```diff
@@ -0,0 +1,543 @@
+struct RequestWrapper {
+    body: Vec<u8>,
+    key: String,
+    settings: RequestSettings,
+}
+
+impl RequestWrapper {
+    fn new(req: PartitionInnerBuffer<Vec<u8>, Bytes>, settings: RequestSettings) -> Self {
+        let (body, key) = req.into_parts();
+
+        // TODO: pull the seconds from the last event
+        let filename = {
+            let seconds = Utc::now().format(&settings.time_format);
+
+            if settings.append_uuid {
+                let uuid = Uuid::new_v4();
+                format!("{}-{}", seconds, uuid.to_hyphenated())
+            } else {
+                seconds.to_string()
+            }
+        };
+
+        let key = format!(
+            "{}{}.{}",
+            String::from_utf8_lossy(&key[..]),
+            filename,
+            settings.extension
+        );
+
+        debug!(
+            message = "sending events.",
+            bytes = &field::debug(body.len()),
+            key = &field::debug(&key)
+        );
+
+        Self {
+            body,
+            key,
+            settings,
+        }
+    }
+}
+
+// Settings required to produce a request that do not change per
+// request. All possible values are pre-computed for direct use in
+// producing a request.
+#[derive(Clone, Debug)]
+struct RequestSettings {
+    acl: HeaderValue,
+    content_type: HeaderValue,
+    content_encoding: Option<HeaderValue>,
+    storage_class: HeaderValue,
+    tags: Vec<(HeaderName, HeaderValue)>,
+    extension: String,
+    time_format: String,
+    append_uuid: bool,
+}
+
+impl RequestSettings {
+    fn new(config: &GcsSinkConfig) -> crate::Result<Self> {
+        let acl = HeaderValue::from_str(&to_string(&config.acl)).unwrap();
+        let content_type = HeaderValue::from_str(config.encoding.content_type()).unwrap();
+        let content_encoding = config
+            .compression
+            .content_encoding()
+            .map(|ce| HeaderValue::from_str(&to_string(ce)).unwrap());
+        let storage_class = HeaderValue::from_str(&to_string(&config.storage_class)).unwrap();
+        let tags = config
+            .tags
+            .as_ref()
+            .map(|tags| tags.iter().map(make_header).collect::<Result<Vec<_>, _>>())
+            .unwrap_or(Ok(vec![]))?;
+        let extension = config
+            .filename_extension
+            .clone()
+            .unwrap_or_else(|| config.compression.extension().into());
+        let time_format = config.filename_time_format.clone().unwrap_or("%s".into());
+        let append_uuid = config.filename_append_uuid.unwrap_or(true);
+        Ok(Self {
+            acl,
+            content_type,
+            content_encoding,
+            storage_class,
+            tags,
+            extension,
+            time_format,
+            append_uuid,
+        })
+    }
+}
+
+// Make a header pair from a key-value string pair
+fn make_header(kv: (&String, &String)) -> crate::Result<(HeaderName, HeaderValue)> {
+    let (name, value) = kv;
```

Comment:
I want to say you can do something like:

```rust
fn make_header((name, value): (&String, &String))
```

Reply:
You are correct. Will fix.